### PR TITLE
fuzzy finder showing C(++) macros

### PR DESCRIPTION
**Pull request by romainfrancois**
### Intent

addresses #11981

### Approach

use `CXTranslationUnit_DetailedPreprocessingRecord` to get macro definition. 

### Automated Tests

> Indicate whether this change includes unit tests or integration tests, or link a work item or pull request to add those tests to another repo. If the change cannot or will not be covered by a test, indicate why.

### QA Notes

> Add additional information for QA on how to validate the change, paying special attention to the level of risk, adjacent areas that could be affected by the change, and any important contextual information not present in the linked issues. 

### Checklist

- [ ] If this PR adds a new feature, or fixes a bug in a previously released version, it includes an entry in `NEWS.md` 
- [ ] If this PR adds or changes UI, the updated UI meets [accessibility standards](https://github.com/rstudio/rstudio/wiki/Accessibility)
- [ ] A reviewer is assigned to this PR (if unsure who to assign, check Area Owners list)
- [ ] This PR passes all local unit tests

### CODE REVIEW

**Comment by romainfrancois** [comment]
image

⚠️  https://clang.llvm.org/doxygen/group__CINDEX__TRANSLATION__UNIT.html says that about the option: 

> Used to indicate that the parser should construct a "detailed" preprocessing record, including all macro definitions and instantiations.
> Constructing a detailed preprocessing record requires more memory and time to parse, since the information contained in the record is usually not retained. However, it can be useful for applications that require more detailed information about the behavior of the preprocessor.

File: src/cpp/session/modules/clang/DefinitionIndex.cpp [hunk 1]
```diff
@@ -83,7 +83,8 @@ CXChildVisitResult cursorVisitor(CXCursor cxCursor,
 
    // ensure it's a definition with linkage or a typedef
    if ((!cursor.isDefinition() || !cursor.hasLinkage()) &&
-       cursorKind != CXCursor_TypedefDecl)
+       cursorKind != CXCursor_TypedefDecl &&
+       cursorKind != CXCursor_MacroDefinition)
```
Comment (romainfrancois) [inline src/cpp/session/modules/clang/DefinitionIndex.cpp:87]: One thing this should maybe do is attempt to remove `define` that are for header guards. Remove macros when the name ends with `_H` or `_HPP` ?

Comment (DavisVaughan) [reply src/cpp/session/modules/clang/DefinitionIndex.cpp:87]: That sounds like a reasonable heuristic to me

I think most of us use that pattern:
cli - https://github.com/r-lib/cli/blob/c39f7957636286562a78d9e6e63abed038298a1a/src/cleancall.h#L2
dplyr - https://github.com/tidyverse/dplyr/blob/138f9fe5bfa8004fa4e7fb8bbd11a15bfa70fcbe/src/dplyr.h#L2
vctrs - https://github.com/r-lib/vctrs/blob/0a219ba242ee368df63bc3876f06f9a6220e85c0/src/altrep-rle.h#L2

I am sure there are some that don't follow this, but I think that is ok

**Comment by romainfrancois** [comment]
Thanks @gtritchie fixed now. No idea what happened.

File: src/cpp/session/modules/clang/DefinitionIndex.cpp [hunk 2]
```diff
@@ -131,6 +136,18 @@ CXChildVisitResult cursorVisitor(CXCursor cxCursor,
    std::string name = cursor.displayName();
    boost::algorithm::replace_last(name, "()", "");
 
+   // skip macros that are header guards
```
[the remark this answers is not in the document]
Comment (DavisVaughan) [reply src/cpp/session/modules/clang/DefinitionIndex.cpp:139]: cpp11 does the `#pragma once` approach, so I'm not sure skipping the first macro definition would work
https://github.com/r-lib/cpp11/blob/3e23b304c016372a85a29e61f0b02fbb19355bea/inst/include/cpp11/altrep.hpp#L1

Unless `#pragma` counts as "the first macro"?

[the remark this answers is not in the document]
Comment (kevinushey) [reply src/cpp/session/modules/clang/DefinitionIndex.cpp:139]: Unfortunately as far as I can see `#pragma`s are not visible to us with libclang 😞  (unless I'm missing something in the API)

I also can't tell if it's possible to get the text of a macro definition from libclang either.

**Comment by romainfrancois** [comment]
Now considering this is a header guard if the macro is the first thing and it matches the pattern `"_h(?pp)?_*$"` 

We might be able to use `location` to read the file and check previous lines for `#ifndef {name}` but I'm not sure this is worth it or if it even works for unsaved files.

**Review by kevinushey — APPROVED** [review]
LGTM!

**Comment by kevinushey** [comment]
> We might be able to use location to read the file and check previous lines for #ifndef {name} but I'm not sure this is worth it or if it even works for unsaved files.

I think you're right. We can revisit this later if we find out that the indexer is picking up too many macro definitions or something similar later.

**Comment by kevinushey** [comment]
Can you also update `NEWS.md`?

**Pull request merged**